### PR TITLE
#I completion, proper glyph for assemblies and more

### PR DESCRIPTION
**Pull request by vasily-kirichenko**
image

It also restricts file types where `#r` and `#load` completion works to `*.fsx` only.

### CODE REVIEW

**Comment by majocha** [comment]
Sweet! Now only `#I` includes to do :)

**Comment by vasily-kirichenko** [comment]
@majocha It would require getting lexer results for entire file, then collect all `#I`s up to current line...

**Comment by majocha** [comment]
Will it aggregate `#I`s from `#load`ed scripts?

**Comment by vasily-kirichenko** [comment]
@majocha it would require reading all `#load`ed files manually and searching for `#I`. I don't think we should do this.

**Comment by vasily-kirichenko** [comment]
I don't know on what compilation stage `#load`s are expanded.

**Comment by vasily-kirichenko** [comment]
image

**Comment by majocha** [comment]
Yes at some point includes are copied to 'tcConfig.includes'

Drilling up from there got me here
[https://github.com/Microsoft/visualfsharp/blob/db4febf3902a68234b3e1a51b246687bb0de3c9e/src/fsharp/fsc.fs#L215](https://github.com/Microsoft/visualfsharp/blob/db4febf3902a68234b3e1a51b246687bb0de3c9e/src/fsharp/fsc.fs#L215)

**Comment by smoothdeveloper** [comment]
@majocha / @vaskir I've dealt with `#I` in VFPT (for handling gotodef on #load) and #I is frankly a mess, I'm sure implementing it 100% "right" would not be easy.

https://github.com/fsprojects/VisualFSharpPowerTools/blob/ed0903551a211cdff5c69bd099c51700a3aaa385/src/FSharp.Editing/Common/UntypedAstUtils.fs#L645

I don't remember if #I from loaded scripts also impacts the loading script itself.

**Comment by vasily-kirichenko** [comment]
@smoothdeveloper I decided to not use the parser to find hash directives because it'd go to Reactor and would be slow.

File: vsintegration/src/FSharp.Editor/Completion/FileSystemCompletion.fs [hunk 1]
```diff
@@ -15,27 +15,33 @@ open System.Text.RegularExpressions
...
 
-    let getItems(provider: CompletionProvider, document: Document, position: int, allowableExtensions: string[], directiveRegex: Regex) =
+    let private getFileGlyph (extention: string) =
+        match extention with
+        | ".exe" | ".dll" -> Some Glyph.Assembly
+        | _ -> None
+
+    let getItems(provider: CompletionProvider, document: Document, position: int, allowableExtensions: string list, directiveRegex: Regex, searchPaths: string list) =
         asyncMaybe {
+            do! Option.guard (Path.GetExtension document.FilePath = ".fsx")
```
Comment (saul) [inline vsintegration/src/FSharp.Editor/Completion/FileSystemCompletion.fs]: What about .fsscript? Why is this here?

File: vsintegration/src/FSharp.Editor/Completion/FileSystemCompletion.fs [hunk 2]
```diff
@@ -89,38 +93,61 @@ module internal FileSystemCompletion =
...
+            let! text = document.GetTextAsync(ct)
+            let lines = text.Lines
+            let caretLine = text.Lines.GetLinePosition(position).Line
+            return
+                lines
+                |> Seq.filter (fun x -> x.LineNumber <= caretLine)
+                |> Seq.choose (fun line ->
+                    let lineStr = line.ToString().Trim()
+                    // optimization: fail fast if the line does not start with "(optional spaces) #I"
+                    if not (lineStr.StartsWith "#I") then None
```
Comment (saul) [inline vsintegration/src/FSharp.Editor/Completion/FileSystemCompletion.fs]: What about #I directives in comment blocks?

File: vsintegration/src/FSharp.Editor/Completion/FileSystemCompletion.fs [hunk 3]
```diff
@@ -89,38 +93,61 @@ module internal FileSystemCompletion =
...
-    override __.IsInsertionTrigger(text, position, _options) = FileSystemCompletion.isInsertionTrigger(text, position)
- 
-    override __.GetTextChangeAsync(selectedItem, ch, cancellationToken) = 
-        match FileSystemCompletion.getTextChange(selectedItem, ch) with
-        | Some x -> Task.FromResult(Nullable x)
-        | None -> base.GetTextChangeAsync(selectedItem, ch, cancellationToken)
\ No newline at end of file
+type internal IncludeDirectiveCompletionProvider() =
+    // we have to pass an extension that's not met in real life because if we pass empty list, it does not filter at all.
+    inherit HashDirectiveCompletionProvider("""\s*#I\s+(@?"*(?<literal>"[^"]*"?))""", [".impossible_extension"], useIncludeDirectives = false)
```
Comment (saul) [inline vsintegration/src/FSharp.Editor/Completion/FileSystemCompletion.fs]: I know this is very picky, but would it not be better to add an invalid path character as the "allowable extension"? Such as "\x00" or similar.

File: vsintegration/src/FSharp.Editor/Completion/FileSystemCompletion.fs (hunk 2 above)
Comment (smoothdeveloper) [inline vsintegration/src/FSharp.Editor/Completion/FileSystemCompletion.fs]: Completing on those would be a feature? ;)

It should be fine to address changes in another PR if you want to make some.

File: vsintegration/src/FSharp.Editor/Completion/FileSystemCompletion.fs (hunk 3 above)
Comment (majocha) [inline vsintegration/src/FSharp.Editor/Completion/FileSystemCompletion.fs]: Looking at Roslyn implementation, empty [ ] should work, too. If I'm not mistaken.

File: vsintegration/src/FSharp.Editor/Completion/FileSystemCompletion.fs (hunk 2 above)
Comment (majocha) [inline vsintegration/src/FSharp.Editor/Completion/FileSystemCompletion.fs]: I'd say as for now it covers 99% of use cases.
There's marginal stuff left out like the one above and #I includes loaded with #load.

File: vsintegration/src/FSharp.Editor/Completion/FileSystemCompletion.fs (hunk 1 above)
Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/Completion/FileSystemCompletion.fs]: I've never seen a single fsscript in real life, ok I'll add it.

File: vsintegration/src/FSharp.Editor/Completion/FileSystemCompletion.fs (hunk 2 above)
Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/Completion/FileSystemCompletion.fs]: It works inside comment blocks. why not? :) Disabling it requires querying lexer.

File: vsintegration/src/FSharp.Editor/Completion/FileSystemCompletion.fs (hunk 3 above)
Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/Completion/FileSystemCompletion.fs]: @majocha I tried empty array, it does not work (no filtering at all). 

@saul I'll try \x00.
